brainiak/utils/fmrisim.py

```diff
@@ -2134,126 +2701,229 @@ def generate_noise(dimensions,
+        sig_voxel = signal_function[:, voxel_counter]
+        noise_voxel = noise_function[:, voxel_counter]
+        magnitude_voxel = magnitude[voxel_counter]
 
-    ax.set_xlim(0, brain.shape[0])
-    ax.set_ylim(0, brain.shape[1])
-    ax.set_zlim(0, brain.shape[2])
+        # Calculate the scaled time course using the specified method
+        if method == 'SFNR':
 
-    # If a mask is provided then plot this
-    if mask is not None:
-        mask_threshold = np.where(np.abs(mask) > 0)
-        ax.scatter(mask_threshold[0],
-                   mask_threshold[1],
-                   mask_threshold[2],
-                   zdir='z',
-                   c='black',
-                   s=10,
-                   alpha=0.01)
-
-    # Plot the volume
-    ax.scatter(brain_threshold[0],
-               brain_threshold[1],
-               brain_threshold[2],
-               zdir='z',
-               c='red',
-               s=20)
-
-    return ax
+            # How much temporal variation is there, relative to the mean
+            # activity
+            temporal_var = noise_voxel.mean() / noise_dict['sfnr']
+
+            # Multiply the timecourse by the variability metric
+            new_sig = sig_voxel * (temporal_var * magnitude_voxel)
+
+        elif method == 'CNR_Amp/Noise-SD':
+
+            # What is the standard deviation of the noise
+            noise_std = np.std(noise_voxel)
+
+            # Multiply the signal timecourse by the the CNR and noise (
+            # rearranging eq.)
+            new_sig = sig_voxel * (magnitude_voxel * noise_std)
+
+        elif method == 'CNR_Amp2/Noise-Var_dB':
+
+            # Calculate the current signal amplitude (likely to be 1,
+            # but not necessarily)
+            sig_amp = np.max(np.abs(sig_voxel))
+
+            # What is the standard deviation of the noise
+            noise_std = np.std(noise_voxel)
+
+            # Rearrange the equation to compute the size of signal change in
+            #  decibels
+            scale = 10 ** ((magnitude_voxel / sig_amp) + np.log10(noise_std
+                                                                  ** 2))
+            new_sig = sig_voxel * np.sqrt(scale)
+
+        elif method == 'CNR_Signal-SD/Noise-SD':
+
+            # What is the standard deviation of the signal and noise
+            sig_std = np.std(sig_voxel)
+            noise_std = np.std(noise_voxel)
+
+            # Multiply the signal timecourse by the the CNR and noise (
+            # rearranging eq.)
+            new_sig = sig_voxel * (magnitude_voxel * noise_std / sig_std)
+
+        elif method == 'CNR_Signal-Var/Noise-Var_dB':
+            # What is the standard deviation of the signal and noise
+            sig_std = np.std(sig_voxel)
+            noise_std = np.std(noise_voxel)
+
+            # Rearrange the equation to compute the size of signal change in
+            #  decibels
+            scale = 10 ** ((magnitude_voxel / sig_std) + np.log10(noise_std
```

Comment:
The same as above, I think (magnitude_voxel / sig_std) should not be in the power of the exponent.

Reply:
Sounds good, changed it to this:

`scale = (10 ** (magnitude_voxel / 20)) * noise_std / (max_amp * sig_std)`